mldsa/sign.h

```diff
@@ -163,10 +184,22 @@ int crypto_sign_signature_extmu(uint8_t *sig, size_t *siglen,
  *              - size_t ctxlen: length of context string
  *              - const uint8_t *sk: pointer to bit-packed secret key
  *
- * Returns 0 (success) or -1 (context string too long)
+ * Returns 0 (success) or -1 (context string too long OR nonce exhausted)
  **************************************************/
 int crypto_sign(uint8_t *sm, size_t *smlen, const uint8_t *m, size_t mlen,
-                const uint8_t *ctx, size_t ctxlen, const uint8_t *sk);
+                const uint8_t *ctx, size_t ctxlen, const uint8_t *sk)
+__contract__(
+  requires(mlen <= SIZE_MAX - CRYPTO_BYTES)
+  requires(memory_no_alias(sm, CRYPTO_BYTES + mlen))
+  requires(memory_no_alias(smlen, sizeof(size_t)))
+  requires(m == sm || memory_no_alias(m, mlen))
+  requires(memory_no_alias(ctx, ctxlen))
+  requires(memory_no_alias(sk, CRYPTO_SECRETKEYBYTES))
+  assigns(memory_slice(sm, CRYPTO_BYTES + mlen))
+  assigns(object_whole(smlen))
+  ensures((return_value == 0 && *smlen == CRYPTO_BYTES + mlen) ||
+          (return_value == -1 && *smlen == mlen))
```

Comment:
Does it make sense to set smlen=mlen in the case that signing fails? Maybe 0 would be better?

Reply:
That's what the reference code does, so I left it alone for now.

Reply:
I weakened the contract to leave *smlen unspecified in the event of error